### ISSUE TITLE
support for std::shared_ptr<const T>

### ISSUE THREAD

**Issue by reder2000** [eyalz800/zpp_bits#109]
Hi, would it be possible to enable this pattern and generally containers of const ?
cereal or boost accept it, it is quite useful in come situations .

I can overcome the problem with  changing line 3316 (ugly)
auto loaded = access::make_unique<std::remove_const_t<type>>();;

However a question remains: serialization of polymorphic types does not seems to work (build back a shared_ptr<Derived> from shared_ptr<Base>)

**Comment by eyalz800** [comment]
Do you have information on how they support it? Sounds like for loading stuff it will always have to reallocate, removing const sounds like a bad solution

**Comment by reder2000** [comment]
I had the same problem with boost 15 years ago (found a bug in shared_ptr this way) but I honestly don't remember.

Do you also save/load only one occurence of a given shared ptr ?

**Comment by eyalz800** [comment]
Not sure about your question - what do you mean by one occurrence of the shared ptr?

**Comment by reder2000** [comment]
if the same shared_ptr appears twice in the traversal of an object, are its contents serialized only once ?

**Comment by reder2000** [comment]
That defeats the purpose of shared_ptr.
However, it would be nice to have  a part of the tutorial explain how to deal with polymorphism ?

**Comment by eyalz800** [comment]
To have that much overhead for supporting this one usecase for shared ptr was and still is debatable  for me. I see shared ptr as an ownership tool to support holding a shared resource between threads more than a tool to share an object inside a compound - that’s what raw pointers are for, and they support even circular pointing which shared ptr would cause a leak doing.

Regarding your second question - polymorphism was discussed in the variant part of the readme - take a look and see if it fits your needs. You can basically use variant of all your different objects who answer the same interface (regardless of inheritance) and use variant of them then visit the variant and call your function.

**Comment by reder2000** [comment]
Thanks for your answer. This makes your library probably inadequate for 99% of use cases, It's a pity because it looked interesting given its ease of use wrt speed.

**Comment by Traveller23** [comment]
@eyalz800 Perhaps this could be done: The library would allow the user to write some additional simple code in the serialization method to satisfy the requirement that the `shared_ptr` be serialized only once. If the user does not provide the additional code, then this library behaves as before, without any performance loss.
